# ¡Hola  !🙋🏻‍♂️

Te escribe Lisandro Saez, soy revisor de código en Tripleten y tengo el agrado de revisar el proyecto que entregaste.

Para simular la dinámica de un ambiente de trabajo, si veo algún error, en primer instancia solo los señalaré, dándote la oportunidad de encontrarlos y corregirlos por tu cuenta. En un trabajo real, el líder de tu equipo hará una dinámica similar. En caso de que no puedas resolver la tarea, te daré una información más precisa en la próxima revisión.

Encontrarás mis comentarios más abajo - **por favor, no los muevas, no los modifiques ni los borres**.

¿Cómo lo voy a hacer? Voy a leer detenidamente cada una de las implementaciones que has llevado a cabo para cumplir con lo solicitado. Verás los comentarios de esta forma:

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Si todo está perfecto.
</div>


<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Si tu código está bien pero se puede mejorar o hay algún detalle que le hace falta. Se aceptan uno o dos comentarios de este tipo en el borrador, pero si hay más, deberías hacer las correcciones. Es como una tarea de prueba al solicitar un trabajo: muchos pequeños errores pueden hacer que un candidato sea rechazado.
</div>

<div class="alert alert-block alert-danger">

<b>Comentario del revisor</b> <a class="tocSkip"></a>

Si de pronto hace falta algo o existe algún problema con tu código o conclusiones.
</div>

Puedes responderme de esta forma (no te preocupes, no es obligatorio):

<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <a class="tocSkip"></a>

Hola, muchas gracias por tus comentarios y la revisión.
</div>

¡Empecemos!

# Déjame escuchar música

Los datos están almacenados en el archivo `/datasets/music_project_en.csv`.

¡Ahora sí, manos al código!


## Etapa 1. Descripción de los datos <a id='data_review'></a>

Abre los datos y examínalos.

Etapa 1.1. Necesitarás `pandas`, así que impórtalo.

In [1]:
import pandas as pd
df = pd.read_csv('/datasets/music_project_en.csv')
print(df.head())


     userID                        Track            artist  genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile   rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg   rock   
2    20EC38            Funiculì funiculà       Mario Lanza    pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice   folk   
4  E2DC1FAE                  Soul People        Space Echo  dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  



<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Bien hecho! Siempre es importante que pasemos el set de datos que estamos usando a `DataFrame`!</div>





<details>
<summary>Haz clic para ver la pista</summary>

Para usar `pandas`, necesitas la palabra clave `import` seguida del nombre de la biblioteca y un alias.


Etapa 1.2. Lee el archivo `music_project_en.csv` de la carpeta `/datasets/` y guárdalo en la variable `df`:

In [2]:
import pandas as pd  # Importa pandas con el alias pd
# Lee el archivo CSV y lo guarda en la variable df
df = pd.read_csv('/datasets/music_project_en.csv')
print(df.head())  # Muestra las primeras filas del DataFrame

     userID                        Track            artist  genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile   rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg   rock   
2    20EC38            Funiculì funiculà       Mario Lanza    pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice   folk   
4  E2DC1FAE                  Soul People        Space Echo  dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  


<details>
<summary>Haz clic para ver la pista</summary>

Usa la palabra clave `read_csv` para leer el archivo CSV. Asegúrate de incluir la ruta `/datasets/music_project_en.csv` y guarda el resultado en la variable `df`.

Etapa 1.3. Muestra las 10 primeras filas de la tabla:

In [3]:
import pandas as pd  # Importa pandas con el alias pd
# Lee el archivo CSV y lo guarda en la variable df
df = pd.read_csv('/datasets/music_project_en.csv')
# Muestra las 10 primeras filas del DataFrame
print(df.head(10))
# filas de la tabla df  ← Esto ya es un comentario válido

     userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5

<details>
<summary>Haz clic para ver la pista</summary>

Usa el método `head` para ver las primeras filas del DataFrame.

Etapa 1.4. Obtén la información general sobre la tabla con el método info().

In [4]:
import pandas as pd  # Importa pandas con el alias pd
# Lee el archivo CSV y lo guarda en la variable df
df = pd.read_csv('/datasets/music_project_en.csv')
# Muestra información general del DataFrame
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB


<details>
<summary>Haz clic para ver la pista</summary>

Usa el método `info` para obtener detalles sobre columnas, tipos de datos y valores nulos.


Estas son nuestras observaciones sobre la tabla. Contiene siete columnas que almacenan los mismos tipos de datos: `object`.

Según la documentación:
- `' userID'`: identificador del usuario;
- `'Track'`: título de la canción;
- `'artist'`: nombre del artista;
- `'genre'`: género de la canción;
- `'City'`: ciudad del usuario;
- `'time'`: la hora exacta en la que se reprodujo la canción;
- `'Day'`: día de la semana.

Podemos ver dos problemas con el estilo en los encabezados de la tabla:
1. Algunos encabezados están en mayúsculas, otros en minúsculas.
2. `Identifica tú mismo el segundo problema y escríbelo aquí.`
Segundo problema de estilo en los encabezados: Los nombres de algunas columnas tienen espacios al principio.
Por ejemplo:
•	' userID' (con un espacio al inicio)
•	' City' (con un espacio al inicio)
Esto es problemático porque puede causar errores o confusión al intentar acceder a esas columnas por nombre, por ejemplo con df['userID'] (fallaría si el nombre real es ' userID' con espacio).

### Escribe algunas observaciones por tu parte. Contesta a las siguientes preguntas: <a id='data_review_conclusions'></a>

`1.   ¿Qué tipo de datos hay en las filas? ¿Cómo podemos saber qué almacenan las columnas?`

`2.   ¿Hay suficientes datos para proporcionar respuestas a nuestra hipótesis o necesitamos más información?`

`3.   ¿Notaste algún problema en los datos, como valores ausentes, duplicados o tipos de datos incorrectos?`

Escribe aquí tus respuestas:

1. Según df.info(), todas las columnas son del tipo object, lo que normalmente indica que contienen texto (strings).

Sin embargo, algunas columnas como 'time' y 'Day' podrían representar fechas y horas, y en ese caso sería mejor convertirlas a tipos más adecuados (datetime, category, etc.).

Podemos usar varias formas para explorar el contenido de cada columna:

df.head() → muestra las primeras filas.
df['nombre_columna'].unique() → muestra los valores únicos.
df['nombre_columna'].value_counts() → cuenta las apariciones de cada valor.
df['nombre_columna'].sample(10) → muestra una muestra aleatoria.
Ejemplo: print(df['genre'].value_counts())

2.Depende de cuál sea tu hipótesis. 
Pero en general: Tienes más de 65,000 registros, lo cual es una buena cantidad de datos.

Los datos incluyen usuarios, canciones, artistas, géneros, ciudades, horarios y días, lo cual es suficiente para analizar tendencias de escucha.

Sin embargo, algunas columnas tienen valores faltantes (como Track, artist, y genre), lo que podría afectar algunos análisis.

3.Sí, se detectan varios problemas:
Valores ausentes (NaN):
Track → faltan 1,343 registros.
artist → faltan 7,567 registros.
genre → faltan 1,198 registros.

Esto se ve en el conteo Non-Null comparado con el total de filas (65,079).
Tipos de datos incorrectos:
time debería ser datetime, no object.
Day podría ser category, ya que contiene días de la semana repetidos.
Tener todo como object limita análisis más avanzados.

Nombres de columnas mal formateados:
Espacios al principio de algunos nombres (' userID', ' City').
Inconsistencia en uso de mayúsculas/minúsculas.


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Gran trabajo! Es importante que antes de encarar cualquier problema que queramos resolver con Python, nos paremos a pensar bien en qué consisten los datos con los que vamos a trabajar!
</div>

## Etapa 2. Preprocesamiento de los datos <a id='data_preprocessing'></a>

Tu objetivo aquí es preparar los datos para analizarlos.
El primer paso es resolver los problemas con los encabezados. Después podemos avanzar a los valores ausentes y duplicados. ¡Empecemos!

Vamos a corregir el formato en los encabezados de la tabla.


### Estilo del encabezado <a id='header_style'></a>
Etapa 2.1. Muestra los encabezados de la tabla (los nombres de las columnas):

In [5]:
# Limpiar encabezados: quitar espacios y convertir a minúsculas
df.columns = df.columns.str.strip().str.lower()
# Verificar los nuevos nombres 
print(df.columns)

Index(['userid', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


<details>
<summary>Haz clic para ver la pista</summary>

Usa el atributo `.columns` para ver todos los nombres de las columnas en el DataFrame.


Vamos cambiar los encabezados de la tabla siguiendo las reglas estilísticas convencionales:
*   Todos los caracteres deben ser minúsculas.
*   Elimina los espacios.
*   Si el nombre tiene varias palabras, utiliza snake_case, es decir, añade un guion bajo ( _ ) entre las palabras en lugar de un espacio.



Etapa 2.2. Utiliza el bucle for para iterar sobre los nombres de las columnas y poner todos los caracteres en minúsculas. Cuando hayas terminado, vuelve a mostrar los encabezados de la tabla:

In [6]:
new_columns = []
for col in df.columns:
    clean_col = col.strip().lower()
    clean_col = clean_col.replace(' ', '_')
    # Ajuste especial para 'userid'
    if clean_col == 'userid':
        clean_col = 'user_id'
    new_columns.append(clean_col)
df.columns = new_columns
print(df.columns)

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


<details>
<summary>Haz clic para ver la pista</summary>

Usa un `for` para recorrer `df.columns` y aplica `.lower()` a cada nombre. Luego, asigna la lista resultante de nuevo a `df.columns`.


Etapa 2.3. Ahora, utilizando el mismo método, elimina los espacios al principio y al final de los nombres de las columnas y muestra los nombres de las columnas de nuevo:

In [7]:
new_columns = []
for col in df.columns:
    clean_col = col.strip()          # Quita espacios al inicio y fin
    clean_col = clean_col.lower()    # Convierte a minúsculas
    clean_col = clean_col.replace(' ', '_')  # Reemplaza espacios internos por guion bajo (snake_case)
    # Ajuste especial para 'userid' a 'user_id'
    if clean_col == 'userid':
        clean_col = 'user_id'
    new_columns.append(clean_col)
df.columns = new_columns
print(df.columns)

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


<details>
<summary>Haz clic para ver la pista</summary>

Usa `.strip()` dentro del bucle para limpiar los nombres de columnas. Asegúrate de reasignar los nuevos nombres a `df.columns`.


Etapa 2.4. Necesitamos aplicar la regla de snake_case en la columna `userid`. Debe ser `user_id`. Cambia el nombre de esta columna y muestra los nombres de todas las columnas cuando hayas terminado.

In [8]:
# Paso 1 a 3: limpiar columnas y estandarizar nombres
new_columns = []
for col in df.columns:
    clean_col = col.strip()
    clean_col = clean_col.lower()
    clean_col = clean_col.replace(' ', '_') 
    new_columns.append(clean_col)
df.columns = new_columns
# Paso 4: renombrar 'userid' a 'user_id' específicamente
df.rename(columns={'userid': 'user_id'}, inplace=True)
# Mostrar nombres finales
print(df.columns)

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


<details>
<summary>Haz clic para ver la pista</summary>

Usa el método `.rename()` con el parámetro `columns` para cambiar `userid` a `user_id`. No olvides establecer `inplace=True` o reasignar a `df`.

Etapa 2.5. Comprueba el resultado. Muestra los encabezados una vez más:

In [9]:
print(df.columns.tolist())
df.rename(columns={' userID': 'user_id'}, inplace=True)
print(df.columns.tolist())

['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day']
['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day']




<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Te felicito, los encabezados se ven geniales!
</div>


<details>
<summary>Haz clic para ver la pista</summary>

Usa `.columns` para volver a ver los nombres actualizados de las columnas.

### Valores ausentes <a id='missing_values'></a>
 Etapa 2.5. Primero, encuentra el número de valores ausentes en la tabla. Debes utilizar dos métodos para obtener el número de valores ausentes.

In [10]:
missing_values = df.isna().sum()
print(missing_values)

user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64


<details>
<summary>Haz clic para ver la pista</summary>

Usa `.isna()` para detectar los valores ausentes y combínalo con el método `.sum()` para contar cuántos hay por columna.


Etapa 2.6. Sustituye los valores ausentes en las columnas `'track'`, `'artist'` y `'genre'` con el string `'unknown'`.

1. Crea una lista llamada columns_to_replace que contenga los nombres de las columnas 'track', 'artist' y 'genre'.

2. Usa un bucle for para iterar sobre cada columna en columns_to_replace.

3. Dentro del bucle, sustituye los valores ausentes en cada columna con el string `'unknown'`.

In [11]:
columns_to_replace = ['track', 'artist', 'genre']
for col in columns_to_replace:
    df[col].fillna('unknown', inplace=True)



<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Te felicito por haber eliminado los NaNs! Es una parte fundamental del análisis de datos!
</div>



<details>
<summary>Haz clic para ver la pista</summary>

Crea una lista con los nombres de las columnas y usa un `for` para recorrerla. Dentro del bucle, aplica `.fillna('unknown')` a cada una.

Etapa 2.7. Ahora comprueba el resultado para asegurarte de que no falten valores ausentes por reemplazar en el conjunto de datos. Para ello, cuenta los valores ausentes una vez más.

In [12]:
missing_after = df.isna().sum()
print(missing_after)

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64


<details>
<summary>Haz clic para ver la pista</summary>

Usa `isna()` con `sum()` otra vez para confirmar que ya no quedan valores nulos en las columnas modificadas.

### Duplicados <a id='duplicates'></a>
Etapa 2.8. Encuentra el número de duplicados explícitos en la tabla. Una vez más, debes aplicar dos métodos para obtener la cantidad de duplicados explícitos.

In [13]:
duplicated_rows = df.duplicated()
num_duplicates = duplicated_rows.sum()
print(f"Número de filas duplicadas explícitas: {num_duplicates}")

Número de filas duplicadas explícitas: 3826


<details>
<summary>Haz clic para ver la pista</summary>

Usa un método que detecte filas repetidas (`duplicated`) y luego otro para contarlas.

Etapa 2.9. Ahora, elimina todos los duplicados. Para ello, llama al método que hace exactamente esto.

In [14]:
df.drop_duplicates(inplace=True)
print(f"Duplicados restantes: {df.duplicated().sum()}")

Duplicados restantes: 0


<details>
<summary>Haz clic para ver la pista</summary>

Busca un método que elimine las filas duplicadas directamente del DataFrame. No olvides actualizar la variable.

Etapa 2.10. Comprobemos ahora si conseguimos eliminar todos los duplicados. Cuenta los duplicados explícitos una vez más para asegurarte de haberlos eliminado todos:

In [15]:
duplicados_restantes = df.duplicated().sum()
print(f"Duplicados restantes: {duplicados_restantes}")

Duplicados restantes: 0



<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Está excelente que te tomes el trabajo extra de revisar que hayan quedado eliminados todos los duplicados!
</div>


<details>
<summary>Haz clic para ver la pista</summary>

Repite el mismo enfoque que usaste para contar duplicados antes. Deberías ver que el total ahora es cero.

Ahora queremos deshacernos de los duplicados implícitos en la columna `genre`. Por ejemplo, el nombre de un género se puede escribir de varias formas. Dichos errores también pueden afectar al resultado.

Etapa 2.11. Primero debemos mostrar una lista de nombres de géneros únicos, por orden alfabético. Para ello:
1. Extrae la columna `genre` del DataFrame.
2. Llama al método que devolverá todos los valores únicos en la columna extraída.


In [16]:
unique_genres = df['genre'].unique()
unique_genres_sorted = sorted(unique_genres)
for genre in unique_genres_sorted:
    print(genre)

acid
acoustic
action
adult
africa
afrikaans
alternative
ambient
americana
animated
anime
arabesk
arabic
arena
argentinetango
art
audiobook
avantgarde
axé
baile
balkan
beats
bigroom
black
bluegrass
blues
bollywood
bossa
brazilian
breakbeat
breaks
broadway
cantautori
cantopop
canzone
caribbean
caucasian
celtic
chamber
children
chill
chinese
choral
christian
christmas
classical
classicmetal
club
colombian
comedy
conjazz
contemporary
country
cuban
dance
dancehall
dancepop
dark
death
deep
deutschrock
deutschspr
dirty
disco
dnb
documentary
downbeat
downtempo
drum
dub
dubstep
eastern
easy
electronic
electropop
emo
entehno
epicmetal
estrada
ethnic
eurofolk
european
experimental
extrememetal
fado
film
fitness
flamenco
folk
folklore
folkmetal
folkrock
folktronica
forró
frankreich
französisch
french
funk
future
gangsta
garage
german
ghazal
gitarre
glitch
gospel
gothic
grime
grunge
gypsy
handsup
hard'n'heavy
hardcore
hardstyle
hardtechno
hip
hip-hop
hiphop
historisch
holiday
hop
horror
house
idm
i

<details>
<summary>Haz clic para ver la pista</summary>

Usa `unique()` para obtener los géneros únicos. Si quieres verlos en orden alfabético, combina con `sorted()`.

Etapa 2.12. Vamos a examinar la lista para identificar **duplicados implícitos** del género `hiphop`, es decir, nombres mal escritos o variantes que hacen referencia al mismo género musical.

Los duplicados que encontrarás son:

* `hip`  
* `hop`  
* `hip-hop`  

Para solucionarlo, vamos a crear una función llamada `replace_wrong_values()`.


1. Define una función llamada `replace_wrong_values()` que reciba los siguientes parámetros:

* `df`: el DataFrame a modificar
* `column`: el nombre de la columna a trabajar
* `wrong_values`: una lista con los valores incorrectos
* `correct_value`: el valor correcto para reemplazar

2. Dentro de la función, usa un bucle `for` para iterar sobre cada valor incorrecto y aplicar `.replace()`.


In [17]:
def replace_wrong_values(df, column, wrong_values, correct_value):
    for wrong in wrong_values:
        df[column].replace(wrong, correct_value, inplace=True)
wrong_genres = ['hip', 'hop', 'hiphop']
replace_wrong_values(df, 'genre', wrong_genres, 'hip-hop')


	
<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Excelente trabajo con los duplicados implícitos!</div>




<details>
<summary>Haz clic para ver la pista</summary>

Crea una función con cuatro parámetros. Dentro del bucle, llama a `.replace()` sobre la columna para cada valor incorrecto, cambiándolo por el correcto.

Etapa 2.13. Ahora, llama a la función pasando:

* `df` como el DataFrame
* `'genre'` como nombre de columna
* `['hip', 'hop', 'hip-hop']` como lista de valores incorrectos
* `'hiphop'` como valor correcto


In [18]:
replace_wrong_values(df, 'genre', ['hip', 'hop', 'hip-hop'], 'hiphop')

<details>
<summary>Haz clic para ver la pista</summary>

Llama a la función que creaste y pasa los cuatro argumentos necesarios. Asegúrate de usar el nombre correcto de la columna y lista los valores que deben unificarse.

Etapa 2.14. Asegúrate de que los nombres duplicados se hayan eliminado. Muestra la lista de valores únicos de la columna `'genre'` una vez más:

In [19]:
unique_genres_cleaned = sorted(df['genre'].unique())
for genre in unique_genres_cleaned:
    print(genre)

acid
acoustic
action
adult
africa
afrikaans
alternative
ambient
americana
animated
anime
arabesk
arabic
arena
argentinetango
art
audiobook
avantgarde
axé
baile
balkan
beats
bigroom
black
bluegrass
blues
bollywood
bossa
brazilian
breakbeat
breaks
broadway
cantautori
cantopop
canzone
caribbean
caucasian
celtic
chamber
children
chill
chinese
choral
christian
christmas
classical
classicmetal
club
colombian
comedy
conjazz
contemporary
country
cuban
dance
dancehall
dancepop
dark
death
deep
deutschrock
deutschspr
dirty
disco
dnb
documentary
downbeat
downtempo
drum
dub
dubstep
eastern
easy
electronic
electropop
emo
entehno
epicmetal
estrada
ethnic
eurofolk
european
experimental
extrememetal
fado
film
fitness
flamenco
folk
folklore
folkmetal
folkrock
folktronica
forró
frankreich
französisch
french
funk
future
gangsta
garage
german
ghazal
gitarre
glitch
gospel
gothic
grime
grunge
gypsy
handsup
hard'n'heavy
hardcore
hardstyle
hardtechno
hiphop
historisch
holiday
horror
house
idm
independent
india

<details>
<summary>Haz clic para ver la pista</summary>

Usa de nuevo el método que muestra los valores únicos de una columna para verificar que todos los nombres se hayan unificado correctamente.

## Etapa 3. Análisis

### Tarea: Comparar el comportamiento de los usuarios en las dos ciudades <a id='activity'></a>

Queremos analizar si hay diferencias en la cantidad de canciones reproducidas en Springfield y Shelbyville. Para ello, usaremos los datos de dos días de la semana: lunes y viernes.

Compararemos cuántas canciones se escucharon en cada ciudad durante esos días para identificar posibles patrones de comportamiento.

Sigue estos tres pasos para organizar tu análisis:

- Dividir: agrupa los datos por ciudad.

- Aplicar: cuenta cuántas canciones se reproducen en cada grupo.

- Combinar: presenta los resultados de forma que se puedan comparar fácilmente ambas ciudades.

Repite este proceso por separado para cada uno de los dos días.

Etapa 3.1.
Cuenta cuántas canciones se reprodujeron en cada ciudad utilizando la columna `'track'` como referencia.

In [20]:
monday_data = df[df['day'].str.lower() == 'monday']
monday_counts = monday_data.groupby('city')['track'].count()
friday_data = df[df['day'].str.lower() == 'friday']
friday_counts = friday_data.groupby('city')['track'].count()
print("Canciones reproducidas el lunes:")
print(monday_counts)
print("\nCanciones reproducidas el viernes:")
print(friday_counts)

Canciones reproducidas el lunes:
city
Shelbyville     5614
Springfield    15740
Name: track, dtype: int64

Canciones reproducidas el viernes:
city
Shelbyville     5895
Springfield    15945
Name: track, dtype: int64



<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Gran trabajo utilizando `groupby()`!</div>


<details>
<summary>Haz clic para ver la pista</summary>

Agrupa los datos por ciudad con el método `groupby()` y usa el método `count()` para contar las canciones reproducidas en cada grupo.

Etapa 3.2. Redacta brevemente tus observaciones sobre los resultados.

¿Qué diferencias encontraste entre Springfield y Shelbyville? ¿A qué podrían deberse?

Escribe tus observaciones aquí.

Etapa 3.3.
Agrupa los datos por día de la semana y cuenta cuántas canciones se reprodujeron los lunes y viernes.



In [21]:
days_of_interest = ['monday', 'friday']
filtered_days = df[df['day'].str.lower().isin(days_of_interest)]
counts_by_day = filtered_days.groupby('day')['track'].count()
print(counts_by_day)

day
Friday    21840
Monday    21354
Name: track, dtype: int64


<details>
<summary>Haz clic para ver la pista</summary>

Cambia el criterio de agrupación a la columna que representa el día de la semana y filtra solo para los días que te interesan.

Etapa 3.4. Describe brevemente qué observaste al comparar los lunes y viernes.

¿Hubo un día con más actividad? ¿Cambia algo si analizas cada ciudad por separado?

Escribe tus observaciones aquí.

Etapa 3.5

Ahora vamos a combinar dos criterios: día y ciudad.

Crea una función llamada `number_tracks()` que reciba dos parámetros:

* `day`: un día de la semana (por ejemplo, `'Monday'`)
* `city`: el nombre de una ciudad (por ejemplo, `'Springfield'`)

Dentro de la función:

1. Filtra el DataFrame por el día.
2. Luego, filtra por la ciudad.
3. Cuenta cuántas veces aparece `'user_id'` en ese filtro.
4. Devuelve ese número como resultado.


In [22]:
# Declara la función number_tracks() con dos parámetros: day y city
def number_tracks(day, city):
    # Almacena las filas del DataFrame donde el valor en la columna 'day' es igual al parámetro day
    day_filter = df['day'].str.lower() == day.lower()
    # Filtra las filas donde el valor en la columna 'city' es igual al parámetro city
    city_filter = df['city'].str.lower() == city.lower()
    # Extrae la columna 'user_id' de la tabla filtrada y aplica el método count()
    count = df[day_filter & city_filter]['user_id'].count()
    # Devuelve el número de valores de la columna 'user_id'
    return count




<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Te felicito por el excelente trabajo que has hecho con las funciones!</div>


<details>
<summary>Haz clic para ver la pista</summary>

Define una función con dos parámetros. Dentro, aplica un filtrado secuencial usando `df[...]`, y al final cuenta los valores con `.count()`.

Etapa 3.6. Llama a `number_tracks()` cuatro veces: una por ciudad en cada uno de los dos días.

In [23]:
def number_tracks(day, city):
    filtered_df = df[(df['day'].str.lower() == day.lower()) & (df['city'].str.lower() == city.lower())]
    return filtered_df['user_id'].count()
# Llama a la función para cada combinación de día y ciudad
count_monday_springfield = number_tracks('Monday', 'Springfield')
# Muestra los resultados
print(f"Canciones reproducidas el lunes en Springfield: {count_monday_springfield}")

Canciones reproducidas el lunes en Springfield: 15740


In [24]:
def number_tracks(day, city):
    filtered_df = df[(df['day'].str.lower() == day.lower()) & (df['city'].str.lower() == city.lower())]
    return filtered_df['user_id'].count()
# Llama a la función para cada combinación de día y ciudad
count_monday_shelbyville = number_tracks('Monday', 'Shelbyville')
# Muestra los resultados
print(f"Canciones reproducidas el lunes en Shelbyville: {count_monday_shelbyville}")

Canciones reproducidas el lunes en Shelbyville: 5614


In [25]:
def number_tracks(day, city):
    filtered_df = df[(df['day'].str.lower() == day.lower()) & (df['city'].str.lower() == city.lower())]
    return filtered_df['user_id'].count()
# Llama a la función para cada combinación de día y ciudad
count_friday_springfield = number_tracks('Friday', 'Springfield')
# Muestra los resultados
print(f"Canciones reproducidas el viernes en Springfield: {count_friday_springfield}")

Canciones reproducidas el viernes en Springfield: 15945


In [26]:
def number_tracks(day, city):
    filtered_df = df[(df['day'].str.lower() == day.lower()) & (df['city'].str.lower() == city.lower())]
    return filtered_df['user_id'].count()
# Llama a la función para cada combinación de día y ciudad
count_friday_shelbyville = number_tracks('Friday', 'Shelbyville')
# Muestra los resultados
print(f"Canciones reproducidas el viernes en Shelbyville: {count_friday_shelbyville}")

Canciones reproducidas el viernes en Shelbyville: 5895


<details>
<summary>Haz clic para ver la pista</summary>

Llama a tu función con combinaciones distintas de ciudad y día, como `'Monday', 'Springfield'`, y guarda o imprime los resultados para comparar.

# Conclusiones <a id='end'></a>

## Escribe tus conclusiones finales sobre el análisis

Redacta un resumen breve y claro de los hallazgos obtenidos durante el proceso de análisis. Tu conclusión debe:

* Mencionar los principales patrones que observaste en los datos.
* Identificar cualquier problema encontrado y cómo lo solucionaste.
* Explicar cómo estas acciones ayudaron a mejorar la calidad del análisis.

Reflexiona también sobre la pregunta central:

> *¿Los datos muestran que el comportamiento de los usuarios —en cuanto a la música que escuchan— varía según la ciudad y el día de la semana?*

Apoya tu respuesta con ejemplos concretos de los resultados obtenidos.
Exploramos los datos de reproducción musical para identificar patrones de comportamiento entre los usuarios de dos ciudades: Springfield y Shelbyville, en dos días específicos de la semana: lunes y viernes.

En general, Springfield registró un mayor número de canciones reproducidas tanto el lunes como el viernes en comparación con Shelbyville.

Si el lunes en Springfield hubo 500 reproducciones y en Shelbyville solo 300, esto sugiere una mayor actividad musical en Springfield al inicio de la semana.

En ambas ciudades, los viernes mostraron un ligero aumento en la cantidad de reproducciones en comparación con los lunes, lo que puede estar relacionado con el inicio del fin de semana y el cambio de estado de ánimo de los usuarios.

Inconsistencias en los nombres de las columnas:Había espacios en blanco y diferencias en el uso de mayúsculas.Se estandarizaron los encabezados aplicando minúsculas, snake_case y eliminando espacios, lo que facilitó la manipulación del DataFrame.

Valores ausentes:Las columnas 'track', 'artist' y 'genre' contenían valores nulos.
Se reemplazaron por el string 'unknown' para evitar errores en el análisis.

Duplicados explícitos e implícitos:Se eliminaron filas completamente duplicadas.
En la columna genre, se unificaron entradas como 'hip', 'hop' y 'hip-hop' en 'hiphop', lo que mejoró la consistencia de los datos categóricos.

¿El comportamiento de los usuarios varía según la ciudad y el día?
Sí, los datos muestran claras variaciones tanto por ubicación geográfica como por día de la semana.

Mayor número de reproducciones en Springfield que en Shelbyville un mismo día indica una posible diferencia en tamaño de población, hábitos o acceso al servicio.

Mayor actividad los viernes en ambas ciudades sugiere un patrón común de consumo musical relacionado con el comportamiento semanal de los usuarios.

Detalla aquí tus conclusiones 
Gracias a la limpieza de datos y análisis estructurado, pudimos identificar patrones claros de uso y diferencias de comportamiento entre ciudades y días. 

Esto puede ser útil para servicios de streaming que busquen personalizar recomendaciones o ajustar estrategias de marketing según el día y ubicación del usuario.